src/ff/galois_field.rs

```diff
@@ -171,12 +172,17 @@ macro_rules! bit_array_impl {
                 type Output = Self;
                 fn mul(self, rhs: Self) -> Self::Output {
                     debug_assert!(2 * $bits < u128::BITS);
-                    let a = <Self as GaloisField>::as_u128(self);
-                    let mut product = 0;
-                    for i in 0..Self::BITS {
-                        let bit = u128::from(rhs[i]);
-                        product ^= bit * (a << i);
-                    }
+                    let a = _mm_set_epi64x(
```

Comment:
Could be wrong but I think this wouldn't work on ARM because it uses different instruction set for carry-less multiplies. Lots of folks use M1 Mac (ARM) so we probably need to support both x86 and ARM and define 3 sets of multiplication instructions: no clmul support (basically what we had before), x86 with clmul support and ARM with clmul support.


In other words, something like this should be enough

```rust
#[cfg(all(
    target_arch = "x86_64",
    target_feature = "pclmulqdq"
))]
fn mul(a: 64, b: 64) -> u128 // or [u64; 2]
{
   ... Intel intrinsics  
}

#[cfg(all(
    target_arch="aarch64",
    target_feature="neon"
)]
fn mul(a: 64, b: 64) -> u128 // or [u64; 2]
{
   ... ARM intrinsics  
}

#[cfg(not(any(all(
    target_arch="aarch64",
    target_feature="neon"), all(
    target_arch = "x86_64",
    target_feature = "pclmulqdq"
)))]
fn mul(a: 64, b: 64) -> u128 // or [u64; 2] 
{
   no hardware support, standard multiply
}

```



some reference materials and I would be happy to help if needed:
https://github.com/rust-lang/stdarch/issues/318
https://github.com/geky/gf256/blob/master/src/xmul.rs